File: src/main/java/com/rtomyj/yugiohAPI/model/banlist/BanListNewContent.java
package com.rtomyj.yugiohAPI.model.banlist;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.rtomyj.yugiohAPI.controller.banlist.BannedCardsController;
import com.rtomyj.yugiohAPI.controller.banlist.BanListNewContentController;
import com.rtomyj.yugiohAPI.controller.banlist.BanListRemovedContentController;

import com.rtomyj.yugiohAPI.helper.constants.SwaggerConstants;
import com.rtomyj.yugiohAPI.model.HateoasLinks;
import io.swagger.annotations.ApiModelProperty;
import lombok.*;
import org.springframework.hateoas.RepresentationModel;

import io.swagger.annotations.ApiModel;

@Data
@EqualsAndHashCode(callSuper=false)
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonPropertyOrder({ "listRequested", "comparedTo", "newCards" })
@ApiModel(
		description = "Cards added to requested ban list that were not in the previous ban list or cards that have a different ban list status (forbidden, limited, semi-limited) compared to the previous ban list."
		, parent = RepresentationModel.class
)
public class BanListNewContent extends RepresentationModel<BanListNewContent> implements HateoasLinks
{

	@ApiModelProperty(
			value = SwaggerConstants.BAN_LIST_START_DATE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String listRequested;

	@ApiModelProperty(
			value = SwaggerConstants.PREVIOUS_BAN_LIST_START_DATE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String comparedTo;

	@ApiModelProperty(
			value = SwaggerConstants.NEWLY_ADDED_CARDS_TO_BAN_LIST_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private NewCards newCards;

	private static final Class<BanListNewContentController> controllerClass = BanListNewContentController.class;
	private static final Class<BannedCardsController> banListController = BannedCardsController.class;
	private static final Class<BanListRemovedContentController> removedController = BanListRemovedContentController.class;


	@Override
	public void setSelfLink()
	{

		this.add(
			linkTo(methodOn(controllerClass).getNewlyAddedContentForBanList(listRequested)).withSelfRel()
		);

	}


	public void setLinks()
	{

		this.setSelfLink();

		this.add(
				linkTo(methodOn(banListController).getBannedCards(listRequested, false, true)).withRel("Ban List Content")
		);
		this.add(
				linkTo(methodOn(removedController).getNewlyRemovedContentForBanList(listRequested)).withRel("Ban List Removed Content")
		);
		this.add(
				linkTo(methodOn(controllerClass).getNewlyAddedContentForBanList(comparedTo)).withRel("Previous Ban Lists' New Content")
		);

		newCards.setLinks();

	}

}

File: src/main/kotlin/com/rtomyj/skc/find/banlist/model/BanListNewContent.kt
package com.rtomyj.skc.banlist.model

import com.fasterxml.jackson.annotation.JsonPropertyOrder
import com.rtomyj.skc.Open
import com.rtomyj.skc.banlist.controller.BanListDiffController
import com.rtomyj.skc.banlist.controller.BannedCardsController
import com.rtomyj.skc.util.constant.SwaggerConstants
import com.rtomyj.skc.util.HateoasLinks
import io.swagger.v3.oas.annotations.media.Schema
import org.springframework.hateoas.RepresentationModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder

@JsonPropertyOrder("listRequested", "comparedTo", "newCards")
@Schema(
    implementation = BanListNewContent::class,
    description = "Cards added to requested ban list that were not in the previous ban list and/or cards that have a different ban list status (forbidden, limited, semi-limited) compared to the previous ban list.",
)
@Open
data class BanListNewContent(
	@Schema(
        implementation = String::class,
        description = SwaggerConstants.BAN_LIST_START_DATE_DESCRIPTION,
    )
    val listRequested: String,

	@Schema(
        implementation = String::class,
        description = SwaggerConstants.PREVIOUS_BAN_LIST_START_DATE_DESCRIPTION,
    )
    val comparedTo: String,

	@Schema(
        implementation = List::class,
        description = "List containing newly forbidden cards and their previous ban status.",
    )
    val newForbidden: List<CardsPreviousBanListStatus>,

	@Schema(
        implementation = List::class,
        description = "List containing newly limited cards and their previous ban status.",
    )
    val newLimited: List<CardsPreviousBanListStatus>,

	@Schema(
        implementation = List::class,
        description = "List containing newly semi-limited cards and their previous ban status.",
    )
    val newSemiLimited: List<CardsPreviousBanListStatus>,

	@Schema(
        implementation = Int::class,
        description = "Total new forbidden cards added to a ban list when compared to a previous logical ban list.",
    )
    val numNewForbidden: Int = newForbidden.size,

	@Schema(
        implementation = Int::class,
        description = "Total new limited cards added to a list when compared to a previous logical ban list.",
    )
    val numNewLimited: Int = newLimited.size,

	@Schema(
        implementation = Int::class,
        description = "Total new semi-limited cards added to a ban list when compared to a previous logical ban list.",
    )
    val numNewSemiLimited: Int = newSemiLimited.size
) : RepresentationModel<BanListNewContent?>(), HateoasLinks {

    companion object {
        private val BAN_LIST_DIFF_CONTROLLER_CLASS = BanListDiffController::class.java
        private val banListController = BannedCardsController::class.java
    }


    override fun setSelfLink() {
        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(BAN_LIST_DIFF_CONTROLLER_CLASS).getNewlyAddedContentForBanList(listRequested)
            ).withSelfRel()
        )
    }


    override fun setLinks() {
        setSelfLink()
        newForbidden
            .forEach(CardsPreviousBanListStatus::setLinks)
        newLimited
            .forEach(CardsPreviousBanListStatus::setLinks)
        newSemiLimited
            .forEach(CardsPreviousBanListStatus::setLinks)

        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(banListController).getBannedCards(listRequested, false, true)
            ).withRel("Ban List Content")
        )
        this.add(
            WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(BAN_LIST_DIFF_CONTROLLER_CLASS)
                    .getNewlyRemovedContentForBanList(listRequested)
            ).withRel("Ban List Removed Content")
        )
    }
}